tensorflow/core/kernels/tensor_forest/prediction_ops.cc

```diff
@@ -0,0 +1,87 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/resource_mgr.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/kernels/tensor_forest/resources.h"
+#include "tensorflow/core/platform/thread_annotations.h"
+#include "tensorflow/core/util/work_sharder.h"
+
+namespace tensorflow {
+
+class TensorForestTreePredictOp : public OpKernel {
+ public:
+  explicit TensorForestTreePredictOp(OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("logits_dimension", &logits_dimension_));
+  }
+
+  void Compute(OpKernelContext* context) override {
+    TensorForestTreeResource* decision_tree_resource;
+    OP_REQUIRES_OK(context, LookupResource(context, HandleFromInput(context, 0),
+                                           &decision_tree_resource));
+    mutex_lock l(*decision_tree_resource->get_mutex());
+    core::ScopedUnref unref_me(decision_tree_resource);
+
+    const Tensor* dense_features = nullptr;
+    OP_REQUIRES_OK(context, context->input("dense_features", &dense_features));
+
+    auto data_set = dense_features->matrix<float>();
+    const int32 batch_size = dense_features->dim_size(0);
+
+    Tensor* output_predictions = nullptr;
+    OP_REQUIRES_OK(context,
+                   context->allocate_output(0, {batch_size, logits_dimension_},
+                                            &output_predictions));
+    auto out = output_predictions->matrix<float>();
+
+    if (decision_tree_resource->get_size() <= 0) {
+      out.setZero();
+      return;
+    }
+    auto* worker_threads = context->device()->tensorflow_cpu_worker_threads();
+    const int32 num_threads = worker_threads->num_threads;
+    const int64 cost_per_traverse = 500;
```

Comment:
Add a todo here to figure out the correct number. This is some magic number that is probably far from the optimal. 
+@yk5 - what your take on this?

Reply:
You need to refer to Shard() for the proper value of this argument.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/work_sharder.h#L34

"cost_per_unit" is an estimate of the number of CPU cycles (or nanoseconds if not CPU-bound) to complete a unit of work.

The value "500" means each Traverse takes 0.5ms on average, IOW, 2000 traverse calls (or 2000 records) are done per second. Is it a fair estimate?

Reply:
it was took from here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/tensor_forest/kernels/model_ops.cc#L207

Reply:
Lets add a TODO here. This cost would probably depend on the depth of the tree we have. We will need to run it on a number of trees of diff depth and see the num of cpu cycles